sycl/source/detail/queue_impl.cpp

```diff
@@ -421,6 +421,116 @@ queue_impl::submit_impl(const detail::type_erased_cgfo_ty &CGF,
 }
 #endif
 
+// no_handler
+
+void queue_impl::extractArgsAndReqsFromLambda(
+    char *LambdaPtr, detail::kernel_param_desc_t (*ParamDescGetter)(int),
+    size_t NumKernelParams, std::vector<ArgDesc> &Args) {
+  size_t IndexShift = 0;
+
+  Args.reserve(NumKernelParams);
+
+  for (size_t I = 0; I < NumKernelParams; ++I) {
+    detail::kernel_param_desc_t ParamDesc = ParamDescGetter(I);
+    void *Ptr = LambdaPtr + ParamDesc.offset;
+    const detail::kernel_param_kind_t &Kind = ParamDesc.kind;
+    const int &Size = ParamDesc.info;
+
+    Args.emplace_back(Kind, Ptr, Size, I + IndexShift);
+  }
+}
+
+void queue_impl::submit_no_handler(
+  const std::shared_ptr<queue_impl> &Self,
+  detail::NDRDescT NDRDesc, const char *KernelName,
```

Comment:
Can we template this on the number of dimensions, like you did for `queue::submit_no_handler`?

I think that would allow us to get rid of `detail::NDRDescT` and just pass the `nd_range<D>` all the way down the stack.  We could get rid of `padId` and `padRange`

Reply:
I don't agree. We have so much instance of ugliest code duplication added with a promise to refactor once design settles. Please do the right thing from the start.